### DISCUSSION TITLE
PWM support

### DISCUSSION THREAD

**Discussion by o-alquimista** [brgl/libgpiod#78] [Q&A]
GPIO Zero has [PWMOutputDevice](https://gpiozero.readthedocs.io/en/stable/api_output.html#pwmoutputdevice), which I use to control the speed of a case fan for the Raspberry Pi 4:

```py
from gpiozero import PWMOutputDevice

fan = PWMOutputDevice(18)
fan.value = 0  # stop (0%)
fan.value = 0.3  # low speed (30%)
fan.value = 0.8  # high speed (80%)
fan.value = 1  # full speed (100%)
```

How can I do the same with libgpiod? `gpioset` only takes 1 or 0 for value.

___

I played with `gpioset --toggle <period> GPIO18=1` to try to emulate slower fan speed by switching it ON and OFF really fast, with periods ranging from 1ms to 100ms. While it did seem to slow down the fan, it didn't operate smoothly — I can hear it stutter (I don't know a better word to describe it).

___

I suppose this is a complicated technique that takes a lot of tweaking. I got a reasonable 50% fan speed with `-t 50ms,10ms`. I'll continue experimenting.

**Comment by brgl** [comment]
libgpiod does not deal with linux PWM, it's a wrapper around the linux GPIO uAPI exclusively.

**Reply by o-alquimista** [reply]
> "Add GPIO PWM driver": https://www.spinics.net/lists/linux-gpio/msg99740.html

Is this related to my issue?

___

So, do you know any "higher-level" library that uses libgpiod?
- GPIO Zero doesn't support libgpiod as pin factory yet, but there's a lot of discussion: https://github.com/gpiozero/gpiozero/issues/840
- WiringPi recently released "New Kernel GPIO device Interface (replaces sysfs functions)": https://github.com/WiringPi/WiringPi/releases/tag/3.4 _(couldn't find out what exactly they're using)_
- [pigpio](http://abyz.me.uk/rpi/pigpio/python.html). If I'm not mistaken, it interacts with GPIO by directly manipulating memory?

**Reply by brgl** [reply]
> > "Add GPIO PWM driver": https://www.spinics.net/lists/linux-gpio/msg99740.html
> 
> Is this related to my issue?
> 

Yes but this is adding a PWM driver not a GPIO driver so libgpiod will not apply.

> So, do you know any "higher-level" library that uses libgpiod?
> 
> * GPIO Zero doesn't support libgpiod as pin factory yet, but there's a lot of discussion: [Support for libgpiod as pin factory? gpiozero/gpiozero#840](https://github.com/gpiozero/gpiozero/issues/840)
> * WiringPi recently released "New Kernel GPIO device Interface (replaces sysfs functions)": https://github.com/WiringPi/WiringPi/releases/tag/3.4 _(couldn't find out what exactly they're using)_
> * [pigpio](http://abyz.me.uk/rpi/pigpio/python.html). If I'm not mistaken, it interacts with GPIO by directly manipulating memory?

I'm not really following what RPi is doing to be honest. This project takes what the mainline linux kernel exposes for GPIO and provides a programming interface to it. It doesn't do much more than that. I don't know what you're trying to do so I can't really advise on how to do it.
